### PR TITLE
Make `dump_to_pickle` great again!

### PR DESCRIPTION
**Pull request by h-mayorquin**
I tesed in core, let's hope it does not break anything else
@DradeAW this should allow you to use `relative_to` with pickle in a fast way. Can you test it for your use case and if it fails show me that json work for that case with a small example?

```python
---------------------
results for branch main
---------------------
---------------------
Pickling with dump_to_pickle
Time elapsed for saving: 0.18831443786621094
Time elapsed for loading: 3.6071527004241943
Memory used by loading: 88.28515625 MiB
---------------------
Pickling directly
Time elapsed for saving: 0.00238800048828125
Time elapsed for loading: 0.14467382431030273
Memory used by loading: 0.625 MiB
Peak memory usage: 285.4609375 MiB

---------------------
results for branch pickle_fix
---------------------
---------------------
Pickling with dump_to_pickle
Time elapsed for saving: 0.06076359748840332
Time elapsed for loading: 0.03839468955993652
Memory used by loading: 0.125 MiB
---------------------
Pickling directly
Time elapsed for saving: 0.0024383068084716797
Time elapsed for loading: 0.03805899620056152
Memory used by loading: 1.125 MiB
Peak memory usage: 194.58203125 MiB

```
Check this script:
https://gist.github.com/h-mayorquin/10fa5b0594d5ce71f0ada1bc7f72f0ec

Main drawback:
The implementation relies on a very weak of metaprograming (we could use global variables as well but I like them even less). Maybe there is a better way of doing this but I just wanted to show a proof of concept.

If this works for @DradeAW I will clean this further.

### CODE REVIEW

File: src/spikeinterface/core/base.py [hunk 1]
```diff
@@ -403,7 +403,7 @@ def to_dict(
         if relative_to is not None:
             relative_to = Path(relative_to).resolve().absolute()
             assert relative_to.is_dir(), "'relative_to' must be an existing directory"
-            dump_dict = _make_paths_relative(dump_dict, relative_to)
+            dump_dict = _make_paths_relative(dump_dict, relative_to, copy=False)
```
Comment (h-mayorquin) [inline src/spikeinterface/core/base.py:406]:
```suggestion
            dump_dict = _make_paths_relative(dump_dict, relative_to)
```

**Comment by samuelgarcia** [comment]
Cool Ramon.
Impressive.
Really nice solution for saving.
Not sure I like the class attribute for loading but this should be easy to go for similar solution.

File: src/spikeinterface/core/base.py [hunk 2]
```diff
@@ -977,16 +1005,16 @@ def save_to_zarr(
...
-def _make_paths_relative(d, relative) -> dict:
+def _make_paths_relative(d, relative, copy=True) -> dict:
     relative = str(Path(relative).resolve().absolute())
     func = lambda p: os.path.relpath(str(p), start=relative)
-    return recursive_path_modifier(d, func, target="path", copy=True)
+    return recursive_path_modifier(d, func, target="path", copy=copy)
 
 
-def _make_paths_absolute(d, base):
+def _make_paths_absolute(d, base, copy=True):
```
Comment (samuelgarcia) [inline src/spikeinterface/core/base.py:1014]: This is a bad idea to expose at top level.
The first call of the recursive must be a copy always to avoid complicated inplace modification.
I do not see where we need copy=False.

Here th logic was copy the first call always and then internally no copy.

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:1014]: I was experimenting with this. Not important to me. Will revert. Sorry for the distraction.

Comment (samuelgarcia) [reply src/spikeinterface/core/base.py:1014]: ok

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py]: Done.

**Comment by DradeAW** [comment]
I find the use of a class variable in this case pretty elegant, good find @h-mayorquin!

I'll test if it works on my hand and get back to you :)

**Comment by samuelgarcia** [comment]
@DradeAW : did you made some tests ?

File: src/spikeinterface/core/base.py [hunk 3]
```diff
@@ -985,7 +1013,7 @@ def _make_paths_relative(d, relative) -> dict:
 
 def _make_paths_absolute(d, base):
     base = Path(base)
-    func = lambda p: str((base / p).resolve().absolute())
+    func = lambda p: str((base / p).resolve())
```
Comment (samuelgarcia) [inline src/spikeinterface/core/base.py:1016]: why ?

Comment (h-mayorquin) [reply src/spikeinterface/core/base.py:1016]: Redundant. Resolve does what absolute does and better:
https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve

**Comment by DradeAW** [comment]
TLDR: It works (can move the folder elsewhere, low RAM usage), but is **extremely slow**!!

Output on `main` with no `relative_to`:
```
**********************************
****** units_categorization ******
**********************************
Loading sortings from previous run...
100%|███████████████████████████████████████████████████████| 9/9 [00:00<00:00, 370.39it/s]


**********************************
******* remove_bad_units_0 *******
**********************************
Loading sortings from previous run...
100%|███████████████████████████████████████████████████████| 9/9 [00:00<00:00, 157.69it/s]


**********************************
******** remove_bad_units ********
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:00<00:00, 55.82it/s]


**********************************
********** align_units ***********
**********************************
Loading sortings from previous run...
100%|███████████████████████████████████████████████████████| 9/9 [00:00<00:00, 193.29it/s]


**********************************
**** remove_duplicated_spikes ****
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:05<00:00,  1.75it/s]


**********************************
******* remove_bad_units_2 *******
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:05<00:00,  1.78it/s]


**********************************
******* remove_bad_units_3 *******
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:05<00:00,  1.77it/s]


**********************************
********** merge_units ***********
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:06<00:00,  1.49it/s]


**********************************
*** remove_duplicated_spikes_2 ***
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:09<00:00,  1.10s/it]


**********************************
******** export_to_phy_0 *********
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:09<00:00,  1.10s/it]
```

Output on this PR with `relative_to`:
```
**********************************
****** units_categorization ******
**********************************
Loading sortings from previous run...
100%|███████████████████████████████████████████████████████| 9/9 [00:00<00:00, 179.13it/s]


**********************************
******* remove_bad_units_0 *******
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:00<00:00, 45.89it/s]


**********************************
******** remove_bad_units ********
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [00:00<00:00, 17.73it/s]


**********************************
********** align_units ***********
**********************************
Loading sortings from previous run...
100%|████████████████████████████████████████████████████████| 9/9 [01:02<00:00,  6.99s/it]


**********************************
**** remove_duplicated_spikes ****
**********************************
Loading sortings from previous run...
 11%|███████████▊                                                                                              | 1/9 [21:44<2:53:57, 1304.63s/it]
% Interrupted because way to slow
```

**Comment by h-mayorquin** [comment]
@DradeAW slow compared to what? Slower than what it was?

Can you also share what you are profiling?

**Comment by DradeAW** [comment]
I run my algorithm lussac (which generated sorting objects that, at each step, are more and more complicated).
I showed the loading time of each sorting at each step (9 sortings), and showed the time for loading on `main` and on this PR.

**Comment by h-mayorquin** [comment]
@DradeAW Btw if you don't have time and this is not that vital for you I am fine with lettings this go. In fact, I kind of prefer to not support `relative_to` to begin with.

**Comment by DradeAW** [comment]
It's not vital, but it would be really convenient to be able to move the processed analyses around.
I'll try to make a simplified case to reproduce this behaviour!

If you want more details, a call would make things easier :)

**Comment by h-mayorquin** [comment]
Thanks for the simplified case, I will save that call for later once I actually understand what's goin on -or I am confused about the right things-
